proposals/p1928.md

```diff
@@ -39,125 +44,142 @@ continuous in memory.
 
-Suggested in [#1787](https://github.com/carbon-language/carbon-lang/pull/1787):
-```carbon
-  var array_1D: [4; i32];
-  var array_2D: [4, 10; i32];
-  var array_nested: [4; [10; i32]];
-```
-This definition is deprived of a defect of previous definition, but the type is
-the last one.
-
-Alternatives:
-```carbon
-  var array_1D[4]: i32;
-  var array_2D[4, 10]: i32;
-  var array_nested[4][10]: i32;
-```
-In this alternative, `auto` will work only for types, not for dimensions of
-array.
-
-```carbon
-  var array_1D: i32[4];
-  var array_2D: i32[4, 10];
-  var array_nested: i32[4][10];
-```
-This definition requires contextual information since `i32[4]` is the same as
-`array_1D[1]` so it is necessary to separate type and variable.
+### Declaration like VLA
 
 Also, it is possible to declare array in function with runtime defined size:
+
 ```carbon
   fn f(n: i32) -> i32 {
     var arr: [n; i32];
     return 0;
   }
 ```
 
-Only of size is defined in compile-time and it is less than
+Only if size is defined in compile-time and it is less than
 `MAX_STACK_SIZE_ARRAY`, array is located in stack memory. Other definitions are
 located in heap memory. It is especially important for the last example.
 
+### Library API
+
+For arrays that size is defined in compile-time, `StaticArray(T, shape)` is.
+`shape` contains information for `dimension_sizes`.
+
+If array size is not defined in compile-time or it should not be allocated in
+stack, `BoxedArray(T, shape)`. Again, `shape` contains information for
+`dimension_sizes`.
+
 ### Indexing
 
-Indexing possible via integer values like in C style:
+Indexing possible by way of integer values like in C style:
+
 ```carbon
   var array: [4; i32];
   var elem: auto = array[1];
 ```
-and also via array:
+
+and also by way of array:
+
 ```carbon
   var array_ind: [3; i32] = (0, 2, 4);
-  var array: [5; i32] = (4, 3, 2, 1, 0);
-  var arrav: auto = array[array_ind];
-  // arrav == (array[0], array[2], array[4]) == (4, 2, 0)
+  var array1: [5; i32] = (4, 3, 2, 1, 0);
+  var array2: auto = array[array_ind];
+  // array2 == (array1[0], array1[2], array1[4]) == (4, 2, 0)
 ```
+
 Array indexing is useful if slices generate arrays.
 
 Indexes are also possible on left-side expression:
+
 ```carbon
   var array: [6; i32] = (5, 4, 3, 2, 1, 0);
   array[(0,1,2)] = array[(3,4,5)]; // array = (2, 1, 0, 2, 1, 0)
 ```
+
 It requires the same sizes on the left and right sides.
 
+at the same time, `array[(0,1)] = array[(1,0)]` will swap elements.
+
 ## Rationale
 
-This proposal mostly has effect on [Software and language evolution](/docs/project/goals.md#software-and-language-evolution).
+-   [Performance-critical software](/docs/project/goals.md#performance-critical-software)
+    -   No need for a lower level language. Developers will use built-in arrays
+        that know enough information itself.
+-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution).
+
+## Alternatives considered
```

Comment:
It's important to explain the reason for not choosing these alternatives, in enough detail that future readers can understand what issues we considered, and why we made the decision we did. In a lot of ways that's going to be the most important part of this proposal. I recommend looking at some of the other documents in `proposals/` to get an idea of what that should look like.